File: sdk/examples/wallet/offline_signing/0_generate_addresses.rs
// Copyright 2022 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

//! In this example we generate addresses which will be used later to find inputs.
//!
//! Rename `.env.example` to `.env` first, then run the command:
//! ```sh
//! cargo run --release --all-features --example 0_generate_addresses
//! ```

use iota_sdk::{
    client::{
        constants::{SHIMMER_BECH32_HRP, SHIMMER_COIN_TYPE},
        secret::{stronghold::StrongholdSecretManager, SecretManager},
    },
    crypto::keys::{bip39::Mnemonic, bip44::Bip44},
    wallet::{ClientOptions, Result, Wallet},
};

const OFFLINE_WALLET_DB_PATH: &str = "./examples/wallet/offline_signing/example-offline-walletdb";
const STRONGHOLD_SNAPSHOT_PATH: &str = "./examples/wallet/offline_signing/example.stronghold";
const ADDRESS_FILE_PATH: &str = "./examples/wallet/offline_signing/example.address.json";

#[tokio::main]
async fn main() -> Result<()> {
    // This example uses secrets in environment variables for simplicity which should not be done in production.
    dotenvy::dotenv().ok();

    let offline_client = ClientOptions::new();

    // Setup Stronghold secret_manager
    let secret_manager = StrongholdSecretManager::builder()
        .password(std::env::var("STRONGHOLD_PASSWORD").unwrap())
        .build(STRONGHOLD_SNAPSHOT_PATH)?;

    let mnemonic = Mnemonic::from(std::env::var("MNEMONIC").unwrap());

    // The mnemonic only needs to be stored the first time
    secret_manager.store_mnemonic(mnemonic).await?;

    // Create the wallet with the secret_manager and client options
    let wallet = Wallet::builder()
        .with_secret_manager(SecretManager::Stronghold(secret_manager))
        .with_storage_path(OFFLINE_WALLET_DB_PATH)
        .with_client_options(offline_client)
        .with_bip_path(Bip44::new(SHIMMER_COIN_TYPE))
        .with_bech32_hrp(SHIMMER_BECH32_HRP)
        .with_alias("Alice")
        .finish()
        .await?;

    println!("Generated a new account '{}'", wallet.alias().await);

    write_addresses_to_file(&wallet).await
}

async fn write_addresses_to_file(wallet: &Wallet) -> Result<()> {
    use tokio::io::AsyncWriteExt;

    let wallet_address = wallet.address().await;
    let json = serde_json::to_string_pretty(&wallet_address)?;
    let mut file = tokio::io::BufWriter::new(tokio::fs::File::create(ADDRESS_FILE_PATH).await?);
    println!("example.address.json:\n{json}");
    file.write_all(json.as_bytes()).await?;
    file.flush().await?;
    Ok(())
}

File: bindings/python/examples/wallet/offline_signing/0_generate_addresses.py
# Copyright 2023 IOTA Stiftung
# SPDX-License-Identifier: Apache-2.0

# In this example we create an account and store its addresses in a file which will be used later to find
# inputs.

import json
import os

from dotenv import load_dotenv

from iota_sdk import ClientOptions, CoinType, StrongholdSecretManager, Wallet, WalletOptions, Bip44

load_dotenv()

OFFLINE_WALLET_DB_PATH = "./wallet/offline_signing/example-offline-walletdb"
STRONGHOLD_SNAPSHOT_PATH = "./wallet/offline_signing/example.stronghold"
ADDRESSES_FILE_PATH = "./wallet/offline_signing/example.addresses.json"


node_url = os.environ.get('NODE_URL', 'https://api.testnet.shimmer.network')
offline_client_options = ClientOptions()

for env_var in ['STRONGHOLD_PASSWORD', 'MNEMONIC']:
    if env_var not in os.environ:
        raise Exception(f'.env {env_var} is undefined, see .env.example')

secret_manager = StrongholdSecretManager(
    STRONGHOLD_SNAPSHOT_PATH, os.environ['STRONGHOLD_PASSWORD'])

bib_path = Bip44(
    coin_type=CoinType.SHIMMER
)

wallet_options = WalletOptions(
    None,
    None,
    bib_path,
    offline_client_options,
    secret_manager,
    OFFLINE_WALLET_DB_PATH)
wallet = Wallet(wallet_options)

# Store the mnemonic in the Stronghold snapshot, this only needs to be
# done once
wallet.store_mnemonic(os.environ['MNEMONIC'])

# Get the address from the wallet
address = wallet.address()

json_data = json.dumps(list(map(lambda x: x.__dict__, address)), indent=4)
print(f"example.addresses.json:\n{json_data}")
f = open(ADDRESSES_FILE_PATH, "w", encoding="utf-8")
f.write(json_data)
f.close()
